File: backend/app/api/v1/endpoints/partner_coupons.py
"""Partner shop coupon CRUD."""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Request

from app.api.utils import success_envelope
from app.api.v1.deps import SessionDep, get_current_partner
from app.schemas.partner_coupon import (
    PartnerCouponCreate,
    PartnerCouponItem,
    PartnerCouponUpdate,
    PartnerCouponValidateRequest,
    PartnerCouponValidateResult,
)
from app.services.partner_coupon_service import PartnerCouponService

router = APIRouter(prefix="/partner/coupons", tags=["partner-coupons"])


@router.get("")
async def list_partner_coupons(
    request: Request,
    session: SessionDep,
    payload: Annotated[dict, Depends(get_current_partner)],
) -> dict:
    rows = await PartnerCouponService(session).list_coupons(UUID(payload["sub"]))
    return success_envelope([PartnerCouponItem.model_validate(r) for r in rows], request)


@router.post("")
async def create_partner_coupon(
    body: PartnerCouponCreate,
    request: Request,
    session: SessionDep,
    payload: Annotated[dict, Depends(get_current_partner)],
) -> dict:
    row = await PartnerCouponService(session).create_coupon(
        UUID(payload["sub"]),
        code=body.code,
        discount_percent=body.discount_percent,
        is_active=body.is_active,
    )
    return success_envelope(PartnerCouponItem.model_validate(row), request)


@router.patch("/{coupon_id}")
async def update_partner_coupon(
    coupon_id: UUID,
    body: PartnerCouponUpdate,
    request: Request,
    session: SessionDep,
    payload: Annotated[dict, Depends(get_current_partner)],
) -> dict:
    row = await PartnerCouponService(session).update_coupon(
        UUID(payload["sub"]),
        coupon_id,
        code=body.code,
        discount_percent=body.discount_percent,
        is_active=body.is_active,
    )
    return success_envelope(PartnerCouponItem.model_validate(row), request)


@router.delete("/{coupon_id}")
async def delete_partner_coupon(
    coupon_id: UUID,
    request: Request,
    session: SessionDep,
    payload: Annotated[dict, Depends(get_current_partner)],
) -> dict:
    await PartnerCouponService(session).delete_coupon(UUID(payload["sub"]), coupon_id)
    return success_envelope({"deleted": True}, request)


@router.post("/validate")
async def validate_partner_coupon(
    body: PartnerCouponValidateRequest,
    request: Request,
    session: SessionDep,
    payload: Annotated[dict, Depends(get_current_partner)],
) -> dict:
    coupon = await PartnerCouponService(session).validate_code(UUID(payload["sub"]), body.code)
    return success_envelope(
        PartnerCouponValidateResult(code=coupon.code, discount_percent=coupon.discount_percent),
        request,
    )

File: frontend/services/partner-coupons.ts
import { api, type ApiEnvelope } from '@/lib/api';

export type PartnerCoupon = {
  id: string;
  code: string;
  discount_percent: string;
  is_active: boolean;
};

export type PartnerCouponValidateResult = {
  code: string;
  discount_percent: string;
};

export async function listPartnerCoupons(): Promise<PartnerCoupon[]> {
  const { data } = await api.get<ApiEnvelope<PartnerCoupon[]>>('/api/v1/partner/coupons');
  return data.data;
}

export async function createPartnerCoupon(input: {
  code: string;
  discount_percent: number;
  is_active?: boolean;
}): Promise<PartnerCoupon> {
  const { data } = await api.post<ApiEnvelope<PartnerCoupon>>('/api/v1/partner/coupons', input);
  return data.data;
}

export async function updatePartnerCoupon(
  id: string,
  input: Partial<{ code: string; discount_percent: number; is_active: boolean }>,
): Promise<PartnerCoupon> {
  const { data } = await api.patch<ApiEnvelope<PartnerCoupon>>(
    `/api/v1/partner/coupons/${id}`,
    input,
  );
  return data.data;
}

export async function deletePartnerCoupon(id: string): Promise<void> {
  await api.delete(`/api/v1/partner/coupons/${id}`);
}

export async function validatePartnerCoupon(code: string): Promise<PartnerCouponValidateResult> {
  const { data } = await api.post<ApiEnvelope<PartnerCouponValidateResult>>(
    '/api/v1/partner/coupons/validate',
    { code },
  );
  return data.data;
}
